src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+		exit(1);
+	}
+
+	if (cap_clear(cap) < 0) {
+		fprintf(stderr, "Could not clear capabilities (cap_clear failed)\n");
+		cap_free(cap);
+		exit(1);
+	}
+
+	for (i = 0; i < 64; i++) {
+		cap_get_flag(cap, i, CAP_EFFECTIVE, &value);
+		if (value == CAP_SET) {
+			fprintf(stderr, "Cannot run irssi with caps set (cap %u)\n", (unsigned)i);
+			cap_free(cap);
+			exit(1);
+		}
+	}
+
+	cap_free(cap);
+#endif
+}
+
+void enforce_resource_limits(void)
+{
+	size_t i;
+	struct rlimit rlim;
+
+	struct rlimit_struct_t {
+		int resource;
+		int limit;
+	};
+	const struct rlimit_struct_t rlimit_struct[9] = {
+		{ RLIMIT_AS,         268435456  }, /* 256 MiB */
+		{ RLIMIT_FSIZE,      1073741824 }, /* 1 GiB */
+		{ RLIMIT_LOCKS,      0          },
+		{ RLIMIT_MEMLOCK,    0          },
+		{ RLIMIT_MSGQUEUE,   0          },
+		{ RLIMIT_NOFILE,     4096       },
+		{ RLIMIT_NPROC,      70         }, /* why is this so large? */
+		{ RLIMIT_SIGPENDING, 0          },
+		{ RLIMIT_STACK,      32768      } /* 32 KiB */
+	};
+	for (i = 0; i < 9; i++) {
+		rlim.rlim_cur = rlimit_struct[i].limit;
+		rlim.rlim_max = rlimit_struct[i].limit;
+
+		if (setrlimit(rlimit_struct[i].resource, &rlim) < 0 && errno != EPERM) {
+			fprintf(stderr, "Could not set resource limits. Errno %d\n", errno);
+			exit(1);
+		}
+	}
+}
+
+void enable_seccomp_sandbox(void)
+{
+	scmp_filter_ctx ctx;
+	size_t i;
+	int rc;
+
+	/* initialize the libseccomp context */
+	ctx = seccomp_init(SCMP_ACT_KILL);
+	if (ctx == NULL) {
+		fprintf(stderr, "Could not initialize the sandbox (seccomp_init failed)\n");
+		exit(1);
+	}
+
+	/* syscalls without argument filtering */
+	const int noarg_whitelist[20] = {
+		SCMP_SYS(brk),
+		SCMP_SYS(chdir),
+		SCMP_SYS(clone),
+		SCMP_SYS(close),
+		SCMP_SYS(exit_group),
+		SCMP_SYS(fstat),
+		SCMP_SYS(getegid),
+		SCMP_SYS(geteuid),
+		SCMP_SYS(getgid),
+		SCMP_SYS(getpid),
+		SCMP_SYS(getuid),
+		SCMP_SYS(lseek),
+		SCMP_SYS(munmap),
+		SCMP_SYS(pipe),
+		SCMP_SYS(rt_sigreturn),
+		SCMP_SYS(set_tid_address),
+		SCMP_SYS(stat),
+		SCMP_SYS(umask),
+		SCMP_SYS(uname)
+	};
+	for (i = 0; i < 20; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, noarg_whitelist[i], 0);
+		if (rc < 0)
+			goto fail;
+
+		rc = seccomp_syscall_priority(ctx, noarg_whitelist[i], 100);
+		if (rc < 0)
+			goto fail;
+	}
+
+	/* mkdir */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mkdir), 1,
+		SCMP_A1(SCMP_CMP_EQ, 0700)); /* src/core/settings.c:828 */
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(mkdir), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* setsockopt */
+	struct setsockopt_struct_t {
+		int sockfd;
+		int level;
+		int optname;
+		socklen_t optlen;
+	};
+	const struct setsockopt_struct_t setsockopt_struct[2] = {
+		{ 4, SOL_SOCKET, SO_KEEPALIVE, 4 },
+		{ 4, SOL_SOCKET, SO_REUSEADDR, 4 }
+	};
+	for (i = 0; i < 2; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(setsockopt), 4,
```

Comment:
Nice work!

Ultimately that would be nice, but you can wait with that change till you're satisfied with the rest of the patchset so as not to disturb the line numbers / comments too much.